physical/consul.go

```diff
@@ -84,14 +160,247 @@ func newConsulBackend(conf map[string]string) (Backend, error) {
 
 	// Setup the backend
 	c := &ConsulBackend{
-		path:       path,
-		client:     client,
-		kv:         client.KV(),
-		permitPool: NewPermitPool(maxParInt),
+		path:                path,
+		client:              client,
+		kv:                  client.KV(),
+		permitPool:          NewPermitPool(maxParInt),
+		consulClientConf:    consulConf,
+		serviceName:         service,
+		checkTimeout:        checkTimeout,
+		checkTimer:          time.NewTimer(checkTimeout),
+		disableRegistration: disableRegistration,
 	}
 	return c, nil
 }
 
+// serviceTags returns all of the relevant tags for Consul.
+func serviceTags(active bool) []string {
+	activeTag := "standby"
+	if active {
+		activeTag = "active"
+	}
+	return []string{activeTag}
+}
+
+func (c *ConsulBackend) AdvertiseActive(active bool) error {
+	c.serviceLock.Lock()
+	defer c.serviceLock.Unlock()
+
+	// Vault is still bootstrapping
+	if c.service == nil {
+		return nil
+	}
+
+	// Save a cached copy of the active state: no way to query Core
+	c.active = active
+
+	// Ensure serial registration to the Consul agent.  Allow for
+	// concurrent calls to update active status while a single task
+	// attempts, until successful, to update the Consul Agent.
+	if !c.disableRegistration && atomic.CompareAndSwapInt64(&c.registrationLock, 0, 1) {
+		defer atomic.CompareAndSwapInt64(&c.registrationLock, 1, 0)
+
+		// Retry agent registration until successful
+		for {
+			c.service.Tags = serviceTags(c.active)
+			agent := c.client.Agent()
+			err := agent.ServiceRegister(c.service)
+			if err == nil {
+				// Success
+				return nil
+			}
+
+			// wtb logger c.logger.Printf("[WARN] service registration failed: %v", err)
+			c.serviceLock.Unlock()
+			time.Sleep(registrationRetryInterval)
+			c.serviceLock.Lock()
+
+			if !c.running {
+				// Shutting down
+				return err
+			}
+		}
+	}
+
+	// Successful concurrent update to active state
+	return nil
+}
+
+func (c *ConsulBackend) AdvertiseSealed(sealed bool) error {
+	c.serviceLock.Lock()
+	defer c.serviceLock.Unlock()
+	c.unsealed = !sealed
+
+	// Vault is still bootstrapping
+	if c.service == nil {
+		return nil
+	}
+
+	if !c.disableRegistration {
+		// Push a TTL check immediately to update the state
+		c.runCheck()
+	}
+
+	return nil
+}
+
+func (c *ConsulBackend) setAdvertiseAddr(addr string) (err error) {
+	c.advertiseHost, c.advertisePort, err = c.parseAdvertiseAddr(addr)
+	if err != nil {
```

Comment:
In fact `setAdvertiseAddr` can be removed since it is only calling a single function.